### PR TITLE
Fix iterator-related memory issues

### LINKED ISSUE #592: [BUG] 'free(): corrupted unsorted chunks' OR 'corrupted double-linked list' fatal errors OR segmentation fault
**Describe the bug**
Memory free or 'corrupted double-linked list' or segmentation fault fatal errors when not consuming the Python generator.

**Steps/Code to reproduce bug**
Running the below snippet:
```
import cucim
slide_path = 'some_svs_file.svs'
level = 0
size = (256, 256)
with cucim.CuImage(slide_path) as slide:
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
```
Results in:
```
free(): corrupted unsorted chunks
Aborted (core dumped)
```
I don't have this issue when I read from the generator, for example:
```
with cucim.CuImage(slide_path) as slide:
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
```
This:
```
with cucim.CuImage(slide_path) as slide:
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
```
Creates a slightly different error:
```
corrupted double-linked list
Aborted (core dumped)
```
This:
```
with cucim.CuImage(slide_path) as slide:
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
    x = slide.read_region([(0, 0), (0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
    x = slide.read_region([(0, 0), (0, 0)], size, level, num_workers=1)
    _ = next(x)
```
Creates yet another error:
```
Segmentation fault (core dumped)
```

**Environment details (please complete the following information):**
 - cucim version: 23.06.00
 - Environment location: Docker
 - Method of cuCIM install: pip

### LINKED ISSUE #598: [BUG] Memory leak in read_region in generator mode (using a list of locations)
**Describe the bug**
I observe a significant memory leak when using cucim `read_region` in generator mode (using a list of locations).

**Steps/Code to reproduce bug**
The following snippet reproduces the memory leak:
```
import cucim
import numpy as np
import random
import os
import psutil

MEMORY_LEAK = True  # Set to False to avoid the memory leak using a workaround...
slide_path = 'some_svs_file.svs'

level = 0
size = (256, 256)
batch_size = 64
num_workers = 16
num_batches = 4.3  # This may be a non-integer number, to have the last iteration return a number of patches < batch_size
locations = [(random.randint(0, 1000), random.randint(2000, 3000)) for _ in range(int(batch_size * num_batches))]
print(f'Number of locations: {len(locations)}, batch size: {batch_size}, number of workers: {num_workers}')

def get_total_mem_usage():
    current_process = psutil.Process(os.getpid())
    mem = current_process.memory_info().rss
    for child in current_process.children(recursive=True):
        mem += child.memory_info().rss
    return mem

with cucim.CuImage(slide_path) as slide:
    start_mem = None
    for i in range(101):
        gen = slide.read_region(locations, size, level, batch_size=batch_size, num_workers=num_workers)
        if MEMORY_LEAK:
            # The regular cucim read_region results in a memory leak when iterating over it using `for x in gen`
            for x in gen:
                _ = np.asarray(x)
        else:
            # When iterating over the read_region generator using `next()`, no memory leak is observed
            num_reads = len(locations) // batch_size + min(1, len(locations) % batch_size)
            for _ in range(num_reads):
                _ = np.asarray(next(gen))

        if i % 10 == 0:
            if start_mem is None:
                start_mem = get_total_mem_usage()
                print(f'starting with: {(start_mem) // (1024 * 1024)} MB of memory consumption')
            else:
                print(f'mem increase (iteration: {i:3d}): {(get_total_mem_usage() - start_mem) // (1024 * 1024):4d} MB')
```
Setting MEMORY_LEAK to False activates a workaround avoiding the memory leak.
Output on my system with the memory leak:
```
Number of locations: 275, batch size: 64, number of workers: 16
starting with: 166 MB of memory consumption
mem increase (iteration:  10):  378 MB
mem increase (iteration:  20):  737 MB
mem increase (iteration:  30): 1099 MB
mem increase (iteration:  40): 1458 MB
mem increase (iteration:  50): 1819 MB
mem increase (iteration:  60): 2178 MB
mem increase (iteration:  70): 2539 MB
mem increase (iteration:  80): 2898 MB
mem increase (iteration:  90): 3259 MB
mem increase (iteration: 100): 3618 MB
```
And without:
```
Number of locations: 275, batch size: 64, number of workers: 16
starting with: 164 MB of memory consumption
mem increase (iteration:  10):   81 MB
mem increase (iteration:  20):   81 MB
mem increase (iteration:  30):   83 MB
mem increase (iteration:  40):   81 MB
mem increase (iteration:  50):   84 MB
mem increase (iteration:  60):   82 MB
mem increase (iteration:  70):   84 MB
mem increase (iteration:  80):   82 MB
mem increase (iteration:  90):   84 MB
mem increase (iteration: 100):   82 MB
```

In the above example, the same slide is read from multiple times.
Note that the issue is also present when iterating the following: reading from a slide and then closing it.
This can be observed by replacing the last part of the above snippet with:
```
start_mem = None
for i in range(101):
    slide = cucim.CuImage(slide_path)
    gen = slide.read_region(locations, size, level, batch_size=batch_size, num_workers=num_workers)
    if MEMORY_LEAK:
        # The regular cucim read_region results in a memory leak when iterating over it using `for x in gen`
        for x in gen:
            _ = np.asarray(x)
    else:
        # When iterating over the read_region generator using `next()`, no memory leak is observed
        num_reads = len(locations) // batch_size + min(1, len(locations) % batch_size)
        for _ in range(num_reads):
            _ = np.asarray(next(gen))

    if i % 10 == 0:
        if start_mem is None:
            start_mem = get_total_mem_usage()
            print(f'starting with: {(start_mem) // (1024 * 1024)} MB of memory consumption')
        else:
            print(f'mem increase (iteration: {i:3d}): {(get_total_mem_usage() - start_mem) // (1024 * 1024):4d} MB')
    slide.close()
```

**Environment details (please complete the following information):**
 - Environment location: Docker, starting from nvcr.io/nvidia/tensorflow:23.06-tf2-py3
 - Method of cuCIM install: pip, v23.06.00
 - CUDA: 12.4, NVidia driver: 525.125.06

### PR DESCRIPTION
**Pull request by gigony**
MERGE after https://github.com/rapidsai/cucim/pull/618

### Fix memory corruption issue in ThreadBatchDataLoader

Fix #592    

- Return a shared pointer of CuImageIterator, taking ownership of CuImageIterator instance, instead of returning a copy of CuImageIterator
- Do not create new instance for CuImageIterator when `__iter__` is called. Instead, return self.
- Wait until all threads are finished in ThreadBatchDataLoader destructor

### Fix memory leak in ThreadBatchDataLoader

Fix #598
    
- Throw StopIteration exception early
- Do not set nullptr for `raster_data_` in `ThreadBatchDataLoader::next_data()`
  - Setting nullptr prevents freeing memory in the destructor of ThreadBatchDataLoader

### Fix fatal errors for cache object

- GIL is not acquired when using Python object.

### CODE REVIEW

File: python/cucim/tests/performance/clara/test_read_region_memory_usage.py [hunk 1]
```diff
@@ -95,3 +95,73 @@ def test_read_random_region_cpu_memleak(testimg_tiff_stripe_4096x4096_256):
 
     # Memory usage difference should be smaller than (iteration) * 100 bytes
     assert mem_usage_history[-1] - mem_usage_history[1] < iteration * 100
+
+
+def test_tiff_iterator(testimg_tiff_stripe_4096x4096_256):
+    """Test if the iterator of read_region does not cause memory leak.
+    From issue #598.
```
Comment (grlee77) [inline python/cucim/tests/performance/clara/test_read_region_memory_usage.py:102]:
```suggestion
    """Verify that the iterator of read_region does not cause a memory leak.

    See issue gh-598: https://github.com/rapidsai/cucim/issues/598
```

File: python/cucim/tests/unit/clara/test_tiff_read_region.py [hunk 2]
```diff
@@ -196,3 +196,40 @@ def test_cuda_array_interface_support(testimg_tiff_stripe_32x24_16_jpeg):
     assert array_interface["version"] == 3
     assert array_interface["mask"] is None
     assert array_interface["stream"] == 1
+
+
+def test_tiff_iterator(testimg_tiff_stripe_4096x4096_256):
+    """Test if the iterator of read_region works well.
+    From issue #592.
```
Comment (grlee77) [inline python/cucim/tests/unit/clara/test_tiff_read_region.py:203]:
```suggestion
    """Test that the iterator of read_region works as expected.

    See issue gh-592: https://github.com/rapidsai/cucim/issues/592
```

**Review by grlee77 — APPROVED** [review]
Thanks, Gigon. It looks good to me. I left an optional suggestion regarding comments on the test cases

**Comment by jakirkham** [comment]
Have merged PR: https://github.com/rapidsai/cucim/pull/618

Looks like there are some conflicts

File: python/cucim/tests/performance/clara/test_read_region_memory_usage.py (hunk 1 above)
Comment (gigony) [reply python/cucim/tests/performance/clara/test_read_region_memory_usage.py:102]: Thanks @grlee77 ! Will update this.

File: python/cucim/tests/unit/clara/test_tiff_read_region.py (hunk 2 above)
Comment (gigony) [reply python/cucim/tests/unit/clara/test_tiff_read_region.py:203]: Thanks @grlee77 for the suggestion!

**Review by jakirkham — APPROVED** [review]

**Comment by gigony** [comment]
I am seeing an error message regarding cpu memory usage:
https://github.com/rapidsai/cucim/actions/runs/6698898284/job/18202421711?pr=620
image

It is likely that it is intermittent issue (may need to increase the threshold > 1273671680 - 1223131136 = 50540544).

**Comment by jakirkham** [comment]
Thanks Gigon! 🙏

Should we file a follow-up issue or PR?

**Comment by jakirkham** [comment]
> It is likely that it is intermittent issue (may need to increase the threshold > 1273671680 - 1223131136 = 50540544).

Filed as issue: https://github.com/rapidsai/cucim/issues/625

**Pull request merged**